diff --git a/pkg/compactor/index_set.go b/pkg/compactor/index_set.go
index 2a850ea616..5eb38607b3 100644
--- a/pkg/compactor/index_set.go
+++ b/pkg/compactor/index_set.go
@@ -10,6 +10,7 @@ import (
 
 	"github.com/go-kit/log"
 	"github.com/go-kit/log/level"
+	"github.com/opentracing/opentracing-go"
 	"github.com/pkg/errors"
 	"github.com/prometheus/prometheus/model/labels"
 
@@ -88,6 +89,9 @@ func newCommonIndexSet(ctx context.Context, tableName string, baseUserIndexSet s
 }
 
 func newIndexSet(ctx context.Context, tableName, userID string, baseIndexSet storage.IndexSet, workingDir string, logger log.Logger) (*indexSet, error) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "compactor.newIndexSet")
+	defer sp.Finish()
+
 	if err := util.EnsureDirectory(workingDir); err != nil {
 		return nil, err
 	}
diff --git a/pkg/compactor/tables_manager.go b/pkg/compactor/tables_manager.go
index a8ad07dff3..249e35d39a 100644
--- a/pkg/compactor/tables_manager.go
+++ b/pkg/compactor/tables_manager.go
@@ -8,6 +8,7 @@ import (
 	"time"
 
 	"github.com/go-kit/log/level"
+	"github.com/opentracing/opentracing-go"
 	"github.com/pkg/errors"
 
 	"github.com/grafana/loki/v3/pkg/compactor/deletion"
@@ -217,6 +218,9 @@ func (c *tablesManager) initTable(ctx context.Context, tableName string) (*table
 }
 
 func (c *tablesManager) runCompaction(ctx context.Context, applyRetention bool) (err error) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, fmt.Sprintf("tablesManager.runCompaction(%v)", applyRetention))
+	defer sp.Finish()
+
 	status := statusSuccess
 	start := time.Now()
 
@@ -320,6 +324,9 @@ func (c *tablesManager) runCompaction(ctx context.Context, applyRetention bool)
 }
 
 func (c *tablesManager) CompactTable(ctx context.Context, tableName string, applyRetention bool) error {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, fmt.Sprintf("tablesManager.CompactTable(%v)", applyRetention))
+	defer sp.Finish()
+
 	schemaCfg, ok := SchemaPeriodForTable(c.schemaConfig, tableName)
 	if !ok {
 		level.Error(util_log.Logger).Log("msg", "skipping compaction since we can't find schema for table", "table", tableName)
diff --git a/pkg/storage/chunk/client/aws/s3_storage_client.go b/pkg/storage/chunk/client/aws/s3_storage_client.go
index 75c885f65c..87f183575e 100644
--- a/pkg/storage/chunk/client/aws/s3_storage_client.go
+++ b/pkg/storage/chunk/client/aws/s3_storage_client.go
@@ -25,6 +25,7 @@ import (
 	"github.com/grafana/dskit/backoff"
 	"github.com/grafana/dskit/flagext"
 	"github.com/grafana/dskit/instrument"
+	"github.com/opentracing/opentracing-go"
 	"github.com/pkg/errors"
 	"github.com/prometheus/client_golang/prometheus"
 	"go.opentelemetry.io/otel"
@@ -516,12 +517,18 @@ func (a *S3ObjectClient) PutObject(ctx context.Context, objectKey string, object
 
 // List implements chunk.ObjectClient.
 func (a *S3ObjectClient) List(ctx context.Context, prefix, delimiter string) ([]client.StorageObject, []client.StorageCommonPrefix, error) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "S3ObjectClient.List")
+	defer sp.Finish()
+
 	var storageObjects []client.StorageObject
 	var commonPrefixes []client.StorageCommonPrefix
 	commonPrefixesSet := make(map[string]bool)
 
 	for i := range a.bucketNames {
 		err := loki_instrument.TimeRequest(ctx, "S3.List", s3RequestDuration, instrument.ErrorCode, func(ctx context.Context) error {
+			sp, ctx := opentracing.StartSpanFromContext(ctx, "S3.List")
+			defer sp.Finish()
+
 			input := s3.ListObjectsV2Input{
 				Bucket:    aws.String(a.bucketNames[i]),
 				Prefix:    aws.String(prefix),
diff --git a/pkg/storage/stores/shipper/indexshipper/downloads/index_set.go b/pkg/storage/stores/shipper/indexshipper/downloads/index_set.go
index 5829ab2490..46ca8727e8 100644
--- a/pkg/storage/stores/shipper/indexshipper/downloads/index_set.go
+++ b/pkg/storage/stores/shipper/indexshipper/downloads/index_set.go
@@ -15,6 +15,7 @@ import (
 	"github.com/go-kit/log"
 	"github.com/go-kit/log/level"
 	"github.com/grafana/dskit/concurrency"
+	"github.com/opentracing/opentracing-go"
 	"golang.org/x/sync/errgroup"
 
 	"github.com/grafana/loki/v3/pkg/storage/chunk/client/util"
@@ -99,6 +100,9 @@ func (t *indexSet) Init(forQuerying bool, logger log.Logger) (err error) {
 	ctx := context.Background()
 	ctx, t.cancelFunc = context.WithTimeout(ctx, downloadTimeout)
 
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "indexSet.Init")
+	defer sp.Finish()
+
 	defer func() {
 		if err != nil {
 			level.Error(logger).Log("msg", "failed to initialize table, cleaning it up", "table", t.tableName, "err", err)
@@ -283,6 +287,9 @@ func (t *indexSet) cleanupDB(fileName string) error {
 }
 
 func (t *indexSet) Sync(ctx context.Context) (err error) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "indexSet.Sync")
+	defer sp.Finish()
+
 	if !t.indexMtx.isReady() {
 		level.Info(t.logger).Log("msg", "skip sync since the index set is not ready")
 		return nil
@@ -293,6 +300,9 @@ func (t *indexSet) Sync(ctx context.Context) (err error) {
 
 // syncWithRetry runs a sync with upto maxSyncRetries on failure
 func (t *indexSet) syncWithRetry(ctx context.Context, lock, bypassListCache bool) error {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "indexSet.syncWithRetry")
+	defer sp.Finish()
+
 	var err error
 	for i := 0; i <= maxSyncRetries; i++ {
 		err = t.sync(ctx, lock, bypassListCache)
@@ -313,6 +323,9 @@ func (t *indexSet) syncWithRetry(ctx context.Context, lock, bypassListCache bool
 
 // sync downloads updated and new files from the storage relevant for the table and removes the deleted ones
 func (t *indexSet) sync(ctx context.Context, lock, bypassListCache bool) (err error) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "indexSet.sync")
+	defer sp.Finish()
+
 	level.Debug(t.logger).Log("msg", fmt.Sprintf("syncing files for table %s", t.tableName))
 
 	toDownload, toDelete, err := t.checkStorageForUpdates(ctx, lock, bypassListCache)
@@ -365,6 +378,9 @@ func (t *indexSet) sync(ctx context.Context, lock, bypassListCache bool) (err er
 
 // checkStorageForUpdates compares files from cache with storage and builds the list of files to be downloaded from storage and to be deleted from cache
 func (t *indexSet) checkStorageForUpdates(ctx context.Context, lock, bypassListCache bool) (toDownload []storage.IndexFile, toDelete []string, err error) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "indexSet.checkStorageForUpdates")
+	defer sp.Finish()
+
 	// listing tables from store
 	var files []storage.IndexFile
 
diff --git a/pkg/storage/stores/shipper/indexshipper/downloads/table.go b/pkg/storage/stores/shipper/indexshipper/downloads/table.go
index 146d89a49c..0987259689 100644
--- a/pkg/storage/stores/shipper/indexshipper/downloads/table.go
+++ b/pkg/storage/stores/shipper/indexshipper/downloads/table.go
@@ -14,6 +14,7 @@ import (
 	"github.com/go-kit/log"
 	"github.com/go-kit/log/level"
 	"github.com/grafana/dskit/concurrency"
+	"github.com/opentracing/opentracing-go"
 	"github.com/pkg/errors"
 	"golang.org/x/sync/errgroup"
 
@@ -270,6 +271,9 @@ func (t *table) DropUnusedIndex(ttl time.Duration, now time.Time) (bool, error)
 
 // Sync downloads updated and new files from the storage relevant for the table and removes the deleted ones
 func (t *table) Sync(ctx context.Context) error {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "table.Sync")
+	defer sp.Finish()
+
 	level.Debug(t.logger).Log("msg", fmt.Sprintf("syncing files for table %s", t.name))
 
 	t.indexSetsMtx.RLock()
diff --git a/pkg/storage/stores/shipper/indexshipper/downloads/table_manager.go b/pkg/storage/stores/shipper/indexshipper/downloads/table_manager.go
index 8e258fb7ae..405b123bb8 100644
--- a/pkg/storage/stores/shipper/indexshipper/downloads/table_manager.go
+++ b/pkg/storage/stores/shipper/indexshipper/downloads/table_manager.go
@@ -13,6 +13,7 @@ import (
 
 	"github.com/go-kit/log"
 	"github.com/go-kit/log/level"
+	"github.com/opentracing/opentracing-go"
 	"github.com/pkg/errors"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/common/model"
@@ -219,6 +220,9 @@ func (tm *tableManager) getOrCreateTable(tableName string) (Table, error) {
 }
 
 func (tm *tableManager) syncTables(ctx context.Context) error {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "tableManager.syncTables")
+	defer sp.Finish()
+
 	tm.tablesMtx.RLock()
 	tables := slices.Collect(maps.Keys(tm.tables))
 	tm.tablesMtx.RUnlock()
diff --git a/pkg/storage/stores/shipper/indexshipper/storage/cached_client.go b/pkg/storage/stores/shipper/indexshipper/storage/cached_client.go
index 6f759b98a1..72bd62c14c 100644
--- a/pkg/storage/stores/shipper/indexshipper/storage/cached_client.go
+++ b/pkg/storage/stores/shipper/indexshipper/storage/cached_client.go
@@ -9,6 +9,7 @@ import (
 	"time"
 
 	"github.com/go-kit/log/level"
+	"github.com/opentracing/opentracing-go"
 	"go.opentelemetry.io/otel/attribute"
 	"go.opentelemetry.io/otel/trace"
 	"golang.org/x/sync/singleflight"
@@ -81,6 +82,9 @@ func (c *cachedObjectClient) RefreshIndexTableCache(ctx context.Context, tableNa
 }
 
 func (c *cachedObjectClient) List(ctx context.Context, prefix, objectDelimiter string, bypassCache bool) ([]client.StorageObject, []client.StorageCommonPrefix, error) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "cachedObjectClient.List")
+	defer sp.Finish()
+
 	if bypassCache {
 		return c.ObjectClient.List(ctx, prefix, objectDelimiter)
 	}
diff --git a/pkg/storage/stores/shipper/indexshipper/storage/client.go b/pkg/storage/stores/shipper/indexshipper/storage/client.go
index 1aba848d8b..790b9e8307 100644
--- a/pkg/storage/stores/shipper/indexshipper/storage/client.go
+++ b/pkg/storage/stores/shipper/indexshipper/storage/client.go
@@ -8,6 +8,7 @@ import (
 	"time"
 
 	"github.com/grafana/loki/v3/pkg/storage/chunk/client"
+	"github.com/opentracing/opentracing-go"
 )
 
 const delimiter = "/"
@@ -63,6 +64,9 @@ func (s *indexStorageClient) RefreshIndexTableCache(ctx context.Context, tableNa
 }
 
 func (s *indexStorageClient) ListTables(ctx context.Context) ([]string, error) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "indexStorageClient.ListTables")
+	defer sp.Finish()
+
 	_, tables, err := s.objectClient.List(ctx, "", delimiter, false)
 	if err != nil {
 		return nil, err
@@ -77,6 +81,9 @@ func (s *indexStorageClient) ListTables(ctx context.Context) ([]string, error) {
 }
 
 func (s *indexStorageClient) ListFiles(ctx context.Context, tableName string, bypassCache bool) ([]IndexFile, []string, error) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "indexStorageClient.ListFiles")
+	defer sp.Finish()
+
 	// The forward slash here needs to stay because we are trying to list contents of a directory without which
 	// we will get the name of the same directory back with hosted object stores.
 	// This is due to the object stores not having a concept of directories.
@@ -106,6 +113,9 @@ func (s *indexStorageClient) ListFiles(ctx context.Context, tableName string, by
 }
 
 func (s *indexStorageClient) ListUserFiles(ctx context.Context, tableName, userID string, bypassCache bool) ([]IndexFile, error) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "indexStorageClient.ListUserFiles")
+	defer sp.Finish()
+
 	// The forward slash here needs to stay because we are trying to list contents of a directory without which
 	// we will get the name of the same directory back with hosted object stores.
 	// This is due to the object stores not having a concept of directories.
diff --git a/pkg/storage/stores/shipper/indexshipper/storage/index_set.go b/pkg/storage/stores/shipper/indexshipper/storage/index_set.go
index 74762a7e3e..2e79ed07fa 100644
--- a/pkg/storage/stores/shipper/indexshipper/storage/index_set.go
+++ b/pkg/storage/stores/shipper/indexshipper/storage/index_set.go
@@ -4,6 +4,8 @@ import (
 	"context"
 	"errors"
 	"io"
+
+	"github.com/opentracing/opentracing-go"
 )
 
 var (
@@ -50,6 +52,9 @@ func (i indexSet) RefreshIndexTableCache(ctx context.Context, tableName string)
 }
 
 func (i indexSet) ListFiles(ctx context.Context, tableName, userID string, bypassCache bool) ([]IndexFile, error) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "indexSet.ListFiles")
+	defer sp.Finish()
+
 	err := i.validateUserID(userID)
 	if err != nil {
 		return nil, err
