diff --git a/internal/component/loki/process/stages/cri.go b/internal/component/loki/process/stages/cri.go
index 2cb82cd0c..5c67e2d99 100644
--- a/internal/component/loki/process/stages/cri.go
+++ b/internal/component/loki/process/stages/cri.go
@@ -19,9 +19,10 @@ import (
 // CRIConfig is an empty struct that is used to enable a pre-defined pipeline
 // for decoding entries that are using the CRI logging format.
 type CRIConfig struct {
-	MaxPartialLines            int    `alloy:"max_partial_lines,attr,optional"`
-	MaxPartialLineSize         uint64 `alloy:"max_partial_line_size,attr,optional"`
-	MaxPartialLineSizeTruncate bool   `alloy:"max_partial_line_size_truncate,attr,optional"`
+	MaxPartialLines            int           `alloy:"max_partial_lines,attr,optional"`
+	MaxPartialLineSize         uint64        `alloy:"max_partial_line_size,attr,optional"`
+	MaxPartialLineSizeTruncate bool          `alloy:"max_partial_line_size_truncate,attr,optional"`
+	MaxPartialLineAge          time.Duration `alloy:"max_partial_line_age,attr,optional"`
 }
 
 var (
@@ -34,6 +35,7 @@ var DefaultCRIConfig = CRIConfig{
 	MaxPartialLines:            100,
 	MaxPartialLineSize:         0,
 	MaxPartialLineSizeTruncate: false,
+	MaxPartialLineAge:          time.Minute,
 }
 
 // SetToDefault implements syntax.Defaulter.
@@ -46,6 +48,9 @@ func (args *CRIConfig) Validate() error {
 	if args.MaxPartialLines <= 0 {
 		return fmt.Errorf("max_partial_lines must be greater than 0")
 	}
+	if args.MaxPartialLineAge <= time.Duration(0) {
+		return fmt.Errorf("max_partial_line_age must be greater than 0")
+	}
 
 	return nil
 }
@@ -96,7 +101,7 @@ const (
 )
 
 func (c *cri) Run(in chan Entry) chan Entry {
-	return RunWithSkipOrSendMany(in, func(e Entry) ([]Entry, bool) {
+	return RunWithSkipOrSendManyWithTick(in, func(e Entry) ([]Entry, bool) {
 		parsed, ok := crip.ParseCRI(e.Line)
 		if !ok {
 			return []Entry{e}, false
@@ -171,6 +176,23 @@ func (c *cri) Run(in chan Entry) chan Entry {
 		}
 
 		return []Entry{e}, false
+	}, 10*time.Second, func() []Entry {
+		// Send partial lines which are left unsent for a while.
+		threshold := time.Now().Add(-c.cfg.MaxPartialLineAge)
+
+		entries := make([]Entry, 0)
+		fingerprints := make([]model.Fingerprint, 0)
+		for k, v := range c.partialLines {
+			if v.Timestamp.Before(threshold) {
+				level.Warn(c.logger).Log("msg", "cri stage: flushing partial line due to max age", "labels", v.Labels)
+				entries = append(entries, v)
+				fingerprints = append(fingerprints, k)
+			}
+		}
+		for _, fp := range fingerprints {
+			delete(c.partialLines, fp)
+		}
+		return entries
 	})
 }
 
diff --git a/internal/component/loki/process/stages/pipeline.go b/internal/component/loki/process/stages/pipeline.go
index 4229cf38d..16e85bafc 100644
--- a/internal/component/loki/process/stages/pipeline.go
+++ b/internal/component/loki/process/stages/pipeline.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"fmt"
 	"sync"
+	"time"
 
 	"github.com/go-kit/log"
 	"github.com/prometheus/client_golang/prometheus"
@@ -170,3 +171,36 @@ func RunWithSkipOrSendMany(input chan Entry, process func(e Entry) ([]Entry, boo
 
 	return out
 }
+
+// RunWithSkipOrSendManyWithTick same as RunWithSkipOrSendMany, except it can run `tick` function periodically.
+func RunWithSkipOrSendManyWithTick(input chan Entry, process func(e Entry) ([]Entry, bool), interval time.Duration, tick func() []Entry) chan Entry {
+	out := make(chan Entry)
+	go func() {
+		defer close(out)
+		ticker := time.NewTicker(interval)
+		defer ticker.Stop()
+		for {
+			select {
+			case e, ok := <-input:
+				if !ok {
+					return
+				}
+				results, skip := process(e)
+				if skip {
+					continue
+				}
+				for _, result := range results {
+					out <- result
+				}
+			case <-ticker.C:
+				results := tick()
+				for _, result := range results {
+					out <- result
+				}
+			}
+		}
+	}()
+
+	return out
+}
+
