diff --git a/pkg/daemon/ceph/client/command.go b/pkg/daemon/ceph/client/command.go
index 32f68e256..ae0e548b9 100644
--- a/pkg/daemon/ceph/client/command.go
+++ b/pkg/daemon/ceph/client/command.go
@@ -41,7 +41,7 @@ const (
 	// CrushTool is the name of the CLI tool for 'crushtool'
 	CrushTool             = "crushtool"
 	CmdExecuteTimeout     = 1 * time.Minute
-	CephConnectionTimeout = "15" // in seconds
+	CephConnectionTimeout = "300" // in seconds
 	// DefaultPGCount will cause Ceph to use the internal default PG count
 	DefaultPGCount = "0"
 )
