diff --git a/controllers/datasource_controller.go b/controllers/datasource_controller.go
index a179790c..7cb57219 100644
--- a/controllers/datasource_controller.go
+++ b/controllers/datasource_controller.go
@@ -275,6 +275,8 @@ func (r *GrafanaDatasourceReconciler) finalize(ctx context.Context, cr *v1beta1.
 }
 
 func (r *GrafanaDatasourceReconciler) onDatasourceCreated(ctx context.Context, grafana *v1beta1.Grafana, cr *v1beta1.GrafanaDatasource, datasource *models.UpdateDataSourceCommand, hash string) error {
+	log := logf.FromContext(ctx)
+
 	if grafana.IsExternal() && cr.Spec.Plugins != nil {
 		return fmt.Errorf("external grafana instances don't support plugins, please remove spec.plugins from your datasource cr")
 	}
@@ -294,6 +296,10 @@ func (r *GrafanaDatasourceReconciler) onDatasourceCreated(ctx context.Context, g
 	}
 
 	if exists && cr.Unchanged(hash) {
+		if found, _ := grafana.Status.Datasources.Find(cr.Namespace, cr.Name); !found {
+			log.Info("fixing inconsistent .status.datasources filed", "grafana", map[string]string{"name": grafana.Name, "namespace": grafana.Namespace})
+			return grafana.AddNamespacedResource(ctx, r.Client, cr, cr.NamespacedResource())
+		}
 		return nil
 	}
 
