diff --git a/processor/tailsamplingprocessor/and_helper.go b/processor/tailsamplingprocessor/and_helper.go
index 239fadfc29..604abe6b55 100644
--- a/processor/tailsamplingprocessor/and_helper.go
+++ b/processor/tailsamplingprocessor/and_helper.go
@@ -25,5 +25,10 @@ func getNewAndPolicy(settings component.TelemetrySettings, config *AndCfg, polic
 
 // Return instance of and sub-policy
 func getAndSubPolicyEvaluator(settings component.TelemetrySettings, cfg *AndSubPolicyCfg, policyExtensions map[string]samplingpolicy.Extension) (samplingpolicy.Evaluator, error) {
-	return getSharedPolicyEvaluator(settings, &cfg.sharedPolicyCfg, policyExtensions)
+	switch cfg.Type {
+	case Nand:
+		return getNewNandPolicy(settings, &cfg.NandCfg, policyExtensions)
+	default:
+		return getSharedPolicyEvaluator(settings, &cfg.sharedPolicyCfg, policyExtensions)
+	}
 }
diff --git a/processor/tailsamplingprocessor/config.go b/processor/tailsamplingprocessor/config.go
index d065ca79cb..df33bf9eb7 100644
--- a/processor/tailsamplingprocessor/config.go
+++ b/processor/tailsamplingprocessor/config.go
@@ -35,6 +35,8 @@ const (
 	And PolicyType = "and"
 	// Not allows defining a Not policy, returning the opposite of the decision of a wrapped policy
 	Not PolicyType = "not"
+	// Nand allows defining a Nand policy, combining the other policies in one and negate
+	Nand PolicyType = "nand"
 	// Drop allows defining a Drop policy, combining one or more policies to drop traces.
 	Drop PolicyType = "drop"
 	// SpanCount sample traces that are have more spans per Trace than a given threshold.
@@ -97,6 +99,9 @@ type CompositeSubPolicyCfg struct {
 // AndSubPolicyCfg holds the common configuration to all policies under and policy.
 type AndSubPolicyCfg struct {
 	sharedPolicyCfg `mapstructure:",squash"` // squash ensures fields are correctly decoded in embedded struct
+
+	// Configs for nand policy evaluator
+	NandCfg NandCfg `mapstructure:"nand"`
 }
 
 // NotSubPolicyCfg holds the common configuration to the policy under the not policy.
@@ -104,6 +109,11 @@ type NotSubPolicyCfg struct {
 	sharedPolicyCfg `mapstructure:",squash"` // squash ensures fields are correctly decoded in embedded struct
 }
 
+// NandSubPolicyCfg holds the common configuration to all policies under nand policy.
+type NandSubPolicyCfg struct {
+	sharedPolicyCfg `mapstructure:",squash"` // squash ensures fields are correctly decoded in embedded struct
+}
+
 // TraceStateCfg holds the common configuration for trace states.
 type TraceStateCfg struct {
 	// Tag that the filter is going to be matching against.
@@ -124,6 +134,13 @@ type NotCfg struct {
 	SubPolicy NotSubPolicyCfg `mapstructure:"not_sub_policy"`
 }
 
+// NandCfg holds the common configuration to all nand policies.
+type NandCfg struct {
+	SubPolicyCfg []NandSubPolicyCfg `mapstructure:"nand_sub_policy"`
+	// prevent unkeyed literal initialization
+	_ struct{}
+}
+
 // DropCfg holds the common configuration to all policies under drop policy.
 type DropCfg struct {
 	SubPolicyCfg []AndSubPolicyCfg `mapstructure:"drop_sub_policy"`
@@ -158,6 +175,8 @@ type PolicyCfg struct {
 	AndCfg AndCfg `mapstructure:"and"`
 	// Configs for defining not policy
 	NotCfg NotCfg `mapstructure:"not"`
+	// Configs for defining nand policy
+	NandCfg NandCfg `mapstructure:"nand"`
 	// Configs for defining drop policy
 	DropCfg DropCfg `mapstructure:"drop"`
 }
diff --git a/processor/tailsamplingprocessor/config_test.go b/processor/tailsamplingprocessor/config_test.go
index 11e9f8ca09..1a3cf3d6d0 100644
--- a/processor/tailsamplingprocessor/config_test.go
+++ b/processor/tailsamplingprocessor/config_test.go
@@ -163,6 +163,30 @@ func TestLoadConfig(t *testing.T) {
 						},
 					},
 				},
+				{
+					sharedPolicyCfg: sharedPolicyCfg{
+						Name: "nand-policy-1",
+						Type: Nand,
+					},
+					NandCfg: NandCfg{
+						SubPolicyCfg: []NandSubPolicyCfg{
+							{
+								sharedPolicyCfg: sharedPolicyCfg{
+									Name:                "test-nand-policy-1",
+									Type:                NumericAttribute,
+									NumericAttributeCfg: NumericAttributeCfg{Key: "key1", MinValue: 50, MaxValue: 100},
+								},
+							},
+							{
+								sharedPolicyCfg: sharedPolicyCfg{
+									Name:               "test-nand-policy-2",
+									Type:               StringAttribute,
+									StringAttributeCfg: StringAttributeCfg{Key: "key2", Values: []string{"value1", "value2"}},
+								},
+							},
+						},
+					},
+				},
 				{
 					sharedPolicyCfg: sharedPolicyCfg{
 						Name: "composite-policy-1",
diff --git a/processor/tailsamplingprocessor/internal/sampling/nand.go b/processor/tailsamplingprocessor/internal/sampling/nand.go
new file mode 100644
index 0000000000..2235ae5ae5
--- /dev/null
+++ b/processor/tailsamplingprocessor/internal/sampling/nand.go
@@ -0,0 +1,45 @@
+// based on and.go
+//
+
+package sampling // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/internal/sampling"
+
+import (
+	"context"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/pkg/samplingpolicy"
+)
+
+type Nand struct {
+	// the subpolicy evaluators
+	subpolicies []samplingpolicy.Evaluator
+	logger      *zap.Logger
+}
+
+func NewNand(
+	logger *zap.Logger,
+	subpolicies []samplingpolicy.Evaluator,
+) samplingpolicy.Evaluator {
+	return &Nand{
+		subpolicies: subpolicies,
+		logger:      logger,
+	}
+}
+
+// Evaluate looks at the trace data and returns a corresponding SamplingDecision.
+func (c *Nand) Evaluate(ctx context.Context, traceID pcommon.TraceID, trace *samplingpolicy.TraceData) (samplingpolicy.Decision, error) {
+	// The policy iterates over all sub-policies and returns NotSampled if all sub-policies returned a Sampled Decision.
+	// If any subpolicy returns NotSampled or InvertNotSampled, it returns Sampled Decision.
+	for _, sub := range c.subpolicies {
+		decision, err := sub.Evaluate(ctx, traceID, trace)
+		if err != nil {
+			return samplingpolicy.Unspecified, err
+		}
+		if decision == samplingpolicy.NotSampled || decision == samplingpolicy.InvertNotSampled {
+			return samplingpolicy.Sampled, nil
+		}
+	}
+	return samplingpolicy.NotSampled, nil
+}
diff --git a/processor/tailsamplingprocessor/internal/sampling/nand_test.go b/processor/tailsamplingprocessor/internal/sampling/nand_test.go
new file mode 100644
index 0000000000..ea946c9c14
--- /dev/null
+++ b/processor/tailsamplingprocessor/internal/sampling/nand_test.go
@@ -0,0 +1,67 @@
+// based on and_test.go
+//
+
+package sampling
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/pkg/samplingpolicy"
+)
+
+func TestNandEvaluatorSampled(t *testing.T) {
+	n1, err := NewStringAttributeFilter(componenttest.NewNopTelemetrySettings(), "name", []string{"value"}, false, 0, false)
+	require.NoError(t, err)
+	n2, err := NewStatusCodeFilter(componenttest.NewNopTelemetrySettings(), []string{"ERROR"})
+	require.NoError(t, err)
+
+	and := NewNand(zap.NewNop(), []samplingpolicy.Evaluator{n1, n2})
+
+	traces := ptrace.NewTraces()
+	rs := traces.ResourceSpans().AppendEmpty()
+	ils := rs.ScopeSpans().AppendEmpty()
+
+	span := ils.Spans().AppendEmpty()
+	span.Status().SetCode(ptrace.StatusCodeError)
+	span.SetTraceID([16]byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16})
+	span.SetSpanID([8]byte{1, 2, 3, 4, 5, 6, 7, 8})
+
+	trace := &samplingpolicy.TraceData{
+		ReceivedBatches: traces,
+	}
+	decision, err := and.Evaluate(t.Context(), traceID, trace)
+	require.NoError(t, err, "Failed to evaluate and policy: %v", err)
+	assert.Equal(t, samplingpolicy.Sampled, decision)
+}
+
+func TestNandEvaluatorNotSampled(t *testing.T) {
+	n1, err := NewStringAttributeFilter(componenttest.NewNopTelemetrySettings(), "attribute_name", []string{"attribute_value"}, false, 0, false)
+	require.NoError(t, err)
+	n2, err := NewStatusCodeFilter(componenttest.NewNopTelemetrySettings(), []string{"ERROR"})
+	require.NoError(t, err)
+
+	and := NewNand(zap.NewNop(), []samplingpolicy.Evaluator{n1, n2})
+
+	traces := ptrace.NewTraces()
+	rs := traces.ResourceSpans().AppendEmpty()
+	ils := rs.ScopeSpans().AppendEmpty()
+
+	span := ils.Spans().AppendEmpty()
+	span.Attributes().PutStr("attribute_name", "attribute_value")
+	span.Status().SetCode(ptrace.StatusCodeError)
+	span.SetTraceID([16]byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16})
+	span.SetSpanID([8]byte{1, 2, 3, 4, 5, 6, 7, 8})
+
+	trace := &samplingpolicy.TraceData{
+		ReceivedBatches: traces,
+	}
+	decision, err := and.Evaluate(t.Context(), traceID, trace)
+	require.NoError(t, err, "Failed to evaluate and policy: %v", err)
+	assert.Equal(t, samplingpolicy.NotSampled, decision)
+}
diff --git a/processor/tailsamplingprocessor/nand_helper.go b/processor/tailsamplingprocessor/nand_helper.go
new file mode 100644
index 0000000000..0d797e8c62
--- /dev/null
+++ b/processor/tailsamplingprocessor/nand_helper.go
@@ -0,0 +1,29 @@
+// based on and_helper.go
+//
+
+package tailsamplingprocessor // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor"
+
+import (
+	"go.opentelemetry.io/collector/component"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/internal/sampling"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/pkg/samplingpolicy"
+)
+
+func getNewNandPolicy(settings component.TelemetrySettings, config *NandCfg, policyExtensions map[string]samplingpolicy.Extension) (samplingpolicy.Evaluator, error) {
+	subPolicyEvaluators := make([]samplingpolicy.Evaluator, len(config.SubPolicyCfg))
+	for i := range config.SubPolicyCfg {
+		policyCfg := &config.SubPolicyCfg[i]
+		policy, err := getNandSubPolicyEvaluator(settings, policyCfg, policyExtensions)
+		if err != nil {
+			return nil, err
+		}
+		subPolicyEvaluators[i] = policy
+	}
+	return sampling.NewNand(settings.Logger, subPolicyEvaluators), nil
+}
+
+// Return instance of nand sub-policy
+func getNandSubPolicyEvaluator(settings component.TelemetrySettings, cfg *NandSubPolicyCfg, policyExtensions map[string]samplingpolicy.Extension) (samplingpolicy.Evaluator, error) {
+	return getSharedPolicyEvaluator(settings, &cfg.sharedPolicyCfg, policyExtensions)
+}
diff --git a/processor/tailsamplingprocessor/nand_helper_test.go b/processor/tailsamplingprocessor/nand_helper_test.go
new file mode 100644
index 0000000000..4ea795d2dc
--- /dev/null
+++ b/processor/tailsamplingprocessor/nand_helper_test.go
@@ -0,0 +1,57 @@
+// based on and_helper_test.go
+//
+
+package tailsamplingprocessor
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/component/componenttest"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/internal/sampling"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/pkg/samplingpolicy"
+)
+
+func TestNandHelper(t *testing.T) {
+	t.Run("valid", func(t *testing.T) {
+		actual, err := getNewNandPolicy(componenttest.NewNopTelemetrySettings(), &NandCfg{
+			SubPolicyCfg: []NandSubPolicyCfg{
+				{
+					sharedPolicyCfg: sharedPolicyCfg{
+						Name:       "test-nand-policy-1",
+						Type:       Latency,
+						LatencyCfg: LatencyCfg{ThresholdMs: 100},
+					},
+				},
+			},
+		}, nil)
+		require.NoError(t, err)
+
+		expected := sampling.NewNand(zap.NewNop(), []samplingpolicy.Evaluator{
+			sampling.NewLatency(componenttest.NewNopTelemetrySettings(), 100, 0),
+		})
+		assert.Equal(t, expected, actual)
+	})
+
+	t.Run("nand under and", func(t *testing.T) {
+		actual, err := getNewAndPolicy(componenttest.NewNopTelemetrySettings(), &AndCfg{
+			SubPolicyCfg: []AndSubPolicyCfg{
+				{
+					sharedPolicyCfg: sharedPolicyCfg{
+						Name: "test-nand-policy-2",
+						Type: Nand,
+					},
+				},
+			},
+		}, nil)
+		require.NoError(t, err)
+
+		expected := sampling.NewAnd(zap.NewNop(), []samplingpolicy.Evaluator{
+			sampling.NewNand(zap.NewNop(), []samplingpolicy.Evaluator{}),
+		})
+		assert.Equal(t, expected, actual)
+	})
+}
diff --git a/processor/tailsamplingprocessor/processor.go b/processor/tailsamplingprocessor/processor.go
index a43aa808ca..5349f823d4 100644
--- a/processor/tailsamplingprocessor/processor.go
+++ b/processor/tailsamplingprocessor/processor.go
@@ -331,6 +331,8 @@ func getPolicyEvaluator(settings component.TelemetrySettings, cfg *PolicyCfg, po
 		return getNewAndPolicy(settings, &cfg.AndCfg, policyExtensions)
 	case Not:
 		return getNewNotPolicy(settings, &cfg.NotCfg, policyExtensions)
+	case Nand:
+		return getNewNandPolicy(settings, &cfg.NandCfg, policyExtensions)
 	case Drop:
 		return getNewDropPolicy(settings, &cfg.DropCfg, policyExtensions)
 	default:
diff --git a/processor/tailsamplingprocessor/testdata/tail_sampling_config.yaml b/processor/tailsamplingprocessor/testdata/tail_sampling_config.yaml
index fc7d219d06..9bb45ba132 100644
--- a/processor/tailsamplingprocessor/testdata/tail_sampling_config.yaml
+++ b/processor/tailsamplingprocessor/testdata/tail_sampling_config.yaml
@@ -106,6 +106,25 @@ tail_sampling:
           }
         }
       },
+      {
+        name: nand-policy-1,
+        type: nand,
+        nand: {
+          nand_sub_policy:
+          [
+            {
+              name: test-nand-policy-1,
+              type: numeric_attribute,
+              numeric_attribute: { key: key1, min_value: 50, max_value: 100 }
+            },
+            {
+                name: test-nand-policy-2,
+                type: string_attribute,
+                string_attribute: { key: key2, values: [ value1, value2 ] }
+            },
+          ]
+        }
+      },
       {
         name: composite-policy-1,
         type: composite,
