diff --git a/pkg/api/accesscontrol.go b/pkg/api/accesscontrol.go
index 2b10380c..6306d3e5 100644
--- a/pkg/api/accesscontrol.go
+++ b/pkg/api/accesscontrol.go
@@ -90,6 +90,9 @@ func FixedRoleRegistrations(viewersCanEdit, dsPermissionsEnforced bool) []ac.Rol
 		Grants: explorerGrants,
 	}
 
+	// Viewers can Explore
+	datasourcesExplorerRole.Grants = append(datasourcesExplorerRole.Grants, string(org.RoleViewer))
+
 	dsReaderGrants := []string{string(org.RoleAdmin)}
 	if !dsPermissionsEnforced {
 		dsReaderGrants = []string{string(org.RoleViewer)}
