diff --git a/debian/rules b/debian/rules
index 1c215a76e50..1c975947507 100755
--- a/debian/rules
+++ b/debian/rules
@@ -26,6 +26,7 @@ extraopts += -DWITH_SYSTEMD=ON -DCEPH_SYSTEMD_ENV_DIR=/etc/default
 extraopts += -DWITH_GRAFANA=ON
 extraopts += -DWITH_RBD_RWL=ON
 extraopts += -DWITH_RBD_SSD_CACHE=ON
+extraopts += -DWITH_ASAN=ON
 # assumes that ceph is exmpt from multiarch support, so we override the libdir.
 extraopts += -DCMAKE_INSTALL_LIBDIR=/usr/lib
 extraopts += -DCMAKE_INSTALL_LIBEXECDIR=/usr/libexec
