diff --git a/debian/rules b/debian/rules
index 9f49deabb0e..f57e74488c8 100755
--- a/debian/rules
+++ b/debian/rules
@@ -39,6 +39,7 @@ else
   extraopts += -DWITH_RBD_RWL=OFF
 endif
 extraopts += -DWITH_RBD_SSD_CACHE=ON
+extraopts += -DWITH_ASAN=ON
 # assumes that ceph is exmpt from multiarch support, so we override the libdir.
 extraopts += -DCMAKE_INSTALL_LIBDIR=/usr/lib
 extraopts += -DCMAKE_INSTALL_LIBEXECDIR=/usr/libexec
