diff --git a/physical/etcd/etcd.go b/physical/etcd/etcd.go
index 985d8def7..37080668c 100644
--- a/physical/etcd/etcd.go
+++ b/physical/etcd/etcd.go
@@ -71,7 +71,8 @@ func NewEtcdBackend(conf map[string]string, logger log.Logger) (physical.Backend
 	case "2", "etcd2", "v2":
 		return newEtcd2Backend(conf, logger)
 	case "3", "etcd3", "v3":
-		if remoteAPIVersion == "2" {
+		// Cybozu: disable auto detection
+		if remoteAPIVersion == "2" && false {
 			return nil, errors.New("etcd3 is required: etcd2 is running")
 		}
 		return newEtcd3Backend(conf, logger)
