diff -ur a/Makefile b/Makefile
--- a/Makefile	2020-04-30 05:07:35.471358583 +0000
+++ b/Makefile	2020-04-30 05:07:45.123293351 +0000
@@ -15,8 +15,7 @@
 	github.com/hashicorp/go-bindata/... \
 	github.com/mitchellh/gox \
 	github.com/kardianos/govendor \
-	github.com/client9/misspell/cmd/misspell \
-	github.com/golangci/golangci-lint/cmd/golangci-lint
+	github.com/client9/misspell/cmd/misspell
 GOFMT_FILES?=$$(find . -name '*.go' | grep -v pb.go | grep -v vendor)
 
 GO_VERSION_MIN=1.12.7
