diff --git a/modules/livestore/instance.go b/modules/livestore/instance.go
index 2bd166c47..77954dcab 100644
--- a/modules/livestore/instance.go
+++ b/modules/livestore/instance.go
@@ -65,6 +65,19 @@ var (
 		Name:      "bytes_received_total",
 		Help:      "The total bytes received per tenant.",
 	}, []string{"tenant", "data_type"})
+	metricCutToWalDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: "tempo_live_store",
+		Name:      "cut_traces_to_wal_duration_seconds",
+		Help:      "Time spent cutting traces to wal",
+		// These buckets are configured to keep the duration shorter than 30s (flush_check_period).
+		Buckets: []float64{
+			1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.0, 4.5,
+			5.0, 6.0, 7.0, 8.0, 9.0,
+			10.0, 11.0, 12.0, 13.0, 14.0, 15.0,
+			16.0, 18.0, 20.0, 22.0, 24.0, 26.0, 28.0,
+			30.0, 35.0, 40.0, 45.0, 50.0, 55.0, 60.0,
+		},
+	}, []string{"tenant", "immediate"})
 	metricBlocksClearedTotal = promauto.NewCounterVec(prometheus.CounterOpts{
 		Namespace: "tempo_live_store",
 		Name:      "blocks_cleared_total",
@@ -136,6 +149,7 @@ type instance struct {
 	// Metrics
 	tracesCreatedTotal prometheus.Counter
 	bytesReceivedTotal *prometheus.CounterVec
+	cutToWalDuration   prometheus.ObserverVec
 
 	overrides overrides.Interface
 }
@@ -162,6 +176,7 @@ func newInstance(instanceID string, cfg Config, wal *wal.WAL, completeBlockEncod
 		overrides:              overrides,
 		tracesCreatedTotal:     metricTracesCreatedTotal.WithLabelValues(instanceID),
 		bytesReceivedTotal:     metricBytesReceivedTotal,
+		cutToWalDuration:       metricCutToWalDuration.MustCurryWith(prometheus.Labels{"tenant": instanceID}),
 		reclaim:                newQuarantine(cfg.BlockReclaimGrace),
 	}
 	i.blocks.Store(emptyBlockSnapshot())
diff --git a/modules/livestore/live_store.go b/modules/livestore/live_store.go
index 51fc3a9ff..0f62f3186 100644
--- a/modules/livestore/live_store.go
+++ b/modules/livestore/live_store.go
@@ -5,6 +5,7 @@ import (
 	"errors"
 	"fmt"
 	"os"
+	"strconv"
 	"sync"
 	"sync/atomic"
 	"time"
@@ -746,6 +747,12 @@ func (s *LiveStore) cutAllInstancesToWal() {
 }
 
 func (s *LiveStore) cutOneInstanceToWal(ctx context.Context, inst *instance, immediate bool) {
+	startTime := time.Now()
+	defer func() {
+		duration := time.Since(startTime).Seconds()
+		inst.cutToWalDuration.WithLabelValues(strconv.FormatBool(immediate)).Observe(duration)
+	}()
+
 	ctx, span := tracer.Start(ctx, "LiveStore.cutOneInstanceToWal",
 		oteltrace.WithAttributes(
 			attribute.String("tenant", inst.tenantID),
