diff --git a/pkg/ingest/config.go b/pkg/ingest/config.go
index c4fb54c82..44ad7bf28 100644
--- a/pkg/ingest/config.go
+++ b/pkg/ingest/config.go
@@ -33,6 +33,9 @@ const (
 	// in the worst case scenario, which is expected to be way above the actual one.
 	maxProducerRecordDataBytesLimit = producerBatchMaxBytes - 16384
 	minProducerRecordDataBytesLimit = 1024 * 1024
+
+	SASLMechanismPlain       = "PLAIN"
+	SASLMechanismScramSHA512 = "SCRAM-SHA-512"
 )
 
 var (
@@ -42,6 +45,8 @@ var (
 	ErrInvalidMaxConsumerLagAtStartup    = errors.New("the configured max consumer lag at startup must greater or equal than the configured target consumer lag")
 	ErrInvalidProducerMaxRecordSizeBytes = fmt.Errorf("the configured producer max record size bytes must be a value between %d and %d", minProducerRecordDataBytesLimit, maxProducerRecordDataBytesLimit)
 	ErrInconsistentSASLCredentials       = errors.New("the SASL username and password must be both configured to enable SASL authentication")
+	ErrInconsistentSASLMechanism         = errors.New("the SASL machanism is specified but the SASL credentials are not specified")
+	ErrInvalidSASLMechanism              = errors.New("the SASL machanism is invalid")
 )
 
 type Config struct {
@@ -64,8 +69,9 @@ type KafkaConfig struct {
 	DialTimeout  time.Duration `yaml:"dial_timeout"`
 	WriteTimeout time.Duration `yaml:"write_timeout"`
 
-	SASLUsername string         `yaml:"sasl_username"`
-	SASLPassword flagext.Secret `yaml:"sasl_password"`
+	SASLUsername  string         `yaml:"sasl_username"`
+	SASLPassword  flagext.Secret `yaml:"sasl_password"`
+	SASLMechanism string         `yaml:"sasl_mechanism"`
 
 	ConsumerGroup                     string        `yaml:"consumer_group"`
 	ConsumerGroupOffsetCommitInterval time.Duration `yaml:"consumer_group_offset_commit_interval"`
@@ -145,6 +151,15 @@ func (cfg *KafkaConfig) Validate() error {
 		return ErrInconsistentSASLCredentials
 	}
 
+	if (cfg.SASLMechanism == "") != (cfg.SASLUsername == "") {
+		return ErrInconsistentSASLMechanism
+	}
+	switch cfg.SASLMechanism {
+	case "", SASLMechanismPlain, SASLMechanismScramSHA512:
+	default:
+		return ErrInvalidSASLMechanism
+	}
+
 	return nil
 }
 
diff --git a/pkg/ingest/writer_client.go b/pkg/ingest/writer_client.go
index 77c27d521..2a4fd3cd3 100644
--- a/pkg/ingest/writer_client.go
+++ b/pkg/ingest/writer_client.go
@@ -14,7 +14,9 @@ import (
 	"github.com/twmb/franz-go/pkg/kerr"
 	"github.com/twmb/franz-go/pkg/kgo"
 	"github.com/twmb/franz-go/pkg/kmsg"
+	"github.com/twmb/franz-go/pkg/sasl"
 	"github.com/twmb/franz-go/pkg/sasl/plain"
+	"github.com/twmb/franz-go/pkg/sasl/scram"
 	"github.com/twmb/franz-go/plugin/kotel"
 	"github.com/twmb/franz-go/plugin/kprom"
 	"go.opentelemetry.io/otel/propagation"
@@ -147,14 +149,26 @@ func commonKafkaClientOptions(cfg KafkaConfig, metrics *kprom.Metrics, logger lo
 		opts = append(opts, kgo.AllowAutoTopicCreation())
 	}
 
-	// SASL plain auth.
+	// SASL PLAIN/SCRAM auth.
 	if cfg.SASLUsername != "" && cfg.SASLPassword.String() != "" {
-		opts = append(opts, kgo.SASL(plain.Plain(func(_ context.Context) (plain.Auth, error) {
-			return plain.Auth{
-				User: cfg.SASLUsername,
-				Pass: cfg.SASLPassword.String(),
-			}, nil
-		})))
+		var m sasl.Mechanism
+		switch cfg.SASLMechanism {
+		case "", SASLMechanismPlain:
+			m = plain.Plain(func(_ context.Context) (plain.Auth, error) {
+				return plain.Auth{
+					User: cfg.SASLUsername,
+					Pass: cfg.SASLPassword.String(),
+				}, nil
+			})
+		case SASLMechanismScramSHA512:
+			m = scram.Sha512(func(_ context.Context) (scram.Auth, error) {
+				return scram.Auth{
+					User: cfg.SASLUsername,
+					Pass: cfg.SASLPassword.String(),
+				}, nil
+			})
+		}
+		opts = append(opts, kgo.SASL(m))
 	}
 
 	tracer := kotel.NewTracer(
